### DISCUSSION TITLE
flux start won't boot on rzansel

### DISCUSSION THREAD

**Discussion by dongahn** [flux-framework/flux-core#3095] [Q&A]
This is w/ the newly tagged versions. 

```
[dahn@rzansel62:flux-sched-0.10.0]$ FLUX_PMI_DEBUG=1 jsrun -a 1 -c ALL_CPUS -g ALL_GPUS --bind=none -n 1 /usr/global/tools/flux/blueos_3_ppc64le_ib/flux-c0.18.0-s0.10.0/bin/flux start  -v hostname
flux-start: /collab/usr/global/tools/flux/blueos_3_ppc64le_ib/flux-c0.18.0-s0.10.0/libexec/flux/cmd/flux-broker hostname
flux-broker: using singleton
2020-07-30T06:37:55.600399Z broker.err[0]: rc1.0: Segmentation fault
2020-07-30T06:37:55.600558Z broker.err[0]: rc1.0: _libcver: Undefined variable.
```

It seems certain environment variables or similar set by jsrun is causing flux broker to segfault.

**Comment by grondo** [comment]
Does `flux start -s 2` work on this platform?

> It seems certain environment variables or similar set by jsrun is causing flux broker to segfault.

I'm not sure it is `flux-broker` that is segfaulting here, but instead the rc1 script or something in it. (the broker would not be able to log `Segmentation fault` if it had experienced the segfault.)

Is there a corefile from which you can determine the actual program that segfaulted?

**Comment by garlick** [comment]
It might also be helpful to add `set -o verbose` to the top of the rc1 script.

**Comment by garlick** [comment]
I'm having trouble getting the pmi _server_ embedded in the shell to emit debug output.  It looks like it should be enabled if we use `flux mini run -o verbose` to set the verbose shell option, but no output for me...

**Comment by grondo** [comment]
I thought those messages were logged at "trace" level,, so try `-o verbose=2`?

**Comment by garlick** [comment]
Oh indeed, that worked.  Sorry for the noise.

**Comment by dongahn** [comment]
I actually got to the bottom of this problem last night but didn't have the time to update the ticket. 

You will be surprised how convoluted this is!

I will update the ticket. In the meantime, please don't spend more time on this.

**Comment by dongahn** [comment]
I found `LD_PRELOAD=/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release/container/../lib/libpami_cudahook.so` set by `jsrun` is what's causing the SEGV.

What's really odd is who is crashing is `/lib64/libc.so.6`.

I believe environment is set up such as way that `/lib64/libc.so.6` is called by some hooks. I don't know what is calling it. My guess is some bash start-up script does this get the libc version.

```
rzansel61{dahn}21: /lib64/libc.so.6
GNU C Library (GNU libc) stable release version 2.17, by Roland McGrath et al.
Copyright (C) 2012 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
Compiled by GNU CC version 4.8.5 20150623 (Red Hat 4.8.5-39).
Compiled on a Linux 3.10.0 system on 2020-01-22.
Available extensions:
	The C stubs add-on version 2.1.2.
	crypt add-on version 2.1 by Michael Glad and others
	GNU Libidn by Simon Josefsson
	Native POSIX Threads Library by Ulrich Drepper et al
	BIND-8.2.3-T5B
	RT using linux kernel aio
libc ABIs: UNIQUE IFUNC
For bug reporting instructions, please see:
<http://www.gnu.org/software/libc/bugs.html>.
```

I am guessing this might be coming from LSF as googling the `_libcver` variable name give some LSF relevant docs.

https://www.ibm.com/support/pages/sourcing-profilelsfcshrclsf-ubuntu-returns-cannot-figure-out-glibc-version

**Comment by dongahn** [comment]
Simplest reproducer:

```console
bash-4.2$ LD_PRELOAD=/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release/container/../lib/libpami_cudahook.so /lib64/libc.so.6
Segmentation fault (core dumped)
```

**Comment by dongahn** [comment]
I will report this to Roy M and if either system admins or IBM can fix it. I am not sure if we can easily work around this by unsetting `LD_PRELOAD`. I can do `jsrun ... bash -c 'unset LD_PRELOAD; flux start ...'` but this isn't pretty.

**Comment by dongahn** [comment]
Two more problems to address to make Flux working most naturally on the new software level on rzansel. It is kind of important to get those problems fixed because this software stack will be deployed Next week Aug 5!

#### jsrun flux start hands with the pmi-shim. Stephen came up with a patch when it was working on Summit.

```diff
diff --git a/src/pmi1.c b/src/pmi1.c
index 2295ad0..e9a3bbf 100644
--- a/src/pmi1.c
+++ b/src/pmi1.c
@@ -60,6 +60,8 @@ PMIX_EXPORT int PMI_Init(int *spawned)
     pmix_info_t info[1];
     bool  val_optinal = 1;
+    //fprintf(stderr, "Initializing PMIx via the pmi-shim\n");
+
     if (PMIX_SUCCESS != (rc = PMIx_Init(&myproc, NULL, 0))) {
         /* if we didn't see a PMIx server (e.g., missing envar),
          * then allow us to run as a singleton */
@@ -175,6 +177,8 @@ PMIX_EXPORT int PMI_KVS_Put(const char kvsname[], const char key[], const char v
         return PMI_SUCCESS;
     }
+    //fprintf(stderr, "Putting %s into %s\n", value, key);
+
     pmix_output_verbose(2, pmix_globals.debug_output,
             "PMI_KVS_Put: KVS=%s, key=%s value=%s", kvsname, key, value);
@@ -257,6 +261,9 @@ PMIX_EXPORT int PMI_KVS_Get( const char kvsname[], const char key[], char value[
     if(sscanf(key, "cmbd.%u.uri", &scanned_rank) > 0) {
         proc.rank = scanned_rank;
         //fprintf(stderr, "Using rank %u for get of %s\n", scanned_rank, key);
+    } else if(strncmp(key, "flux.instance-level", 21) == 0) {
+        //fprintf(stderr, "Error'ing out for get of %s\n", key);
+        return -1;
     } else {
         proc.rank = PMIX_RANK_UNDEF;
         //fprintf(stderr, "Using rank undefined for get of %s\n", key);
@@ -268,6 +275,7 @@ PMIX_EXPORT int PMI_KVS_Get( const char kvsname[], const char key[], char value[
             rc = PMIX_ERROR;
         } else if (NULL != val->data.string) {
             pmix_strncpy(value, val->data.string, length-1);
+            //fprintf(stderr, "Successfully got %s for %s\n", value, key);
         }
         PMIX_VALUE_RELEASE(val);
     }
```

I will clean up this patch and create a new pmi-shim package. I will test if the new pmi-shim works both for Lassen and rzansel.  


#### flux mini run segfault on an application built with the default MPI on rzansel. 

```
/lib64/libSegFault.so(+0x1694)[0x7ffff7f31694]
??:0(??)[0x7ffff7f904d8]
??:0(??)[0x0]
/lib64/libc.so.6(strtoll+0x2c)[0x7ffff7a568dc]
/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release/lib/libmpi_ibm.so.3(ompi_hook_ibmlicense_mpi_init_top+0x98)[0x7ffff7e86a88]
/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release/lib/libmpi_ibm.so.3(ompi_hook_base_mpi_init_top+0x1b4)[0x7ffff7e7fd34]
/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release/lib/libmpi_ibm.so.3(ompi_mpi_init+0x6c)[0x7ffff7dfa0fc]
/usr/tce/packages/spectrum-mpi/ibm/spectrum-mpi-rolling-release/lib/libmpi_ibm.so.3(MPI_Init+0x98)[0x7ffff7e327b8]
/g/g0/dahn/virtual_ring_mpi.c:50(main)[0x1000110c]
/lib64/libc.so.6(+0x25300)[0x7ffff7a35300]
/lib64/libc.so.6(__libc_start_main+0xc4)[0x7ffff7a354f4]
```

This is again the same issue Stephen encountered on Summit. I will report this to Roy. As Veronica from ORNL and we report the same issue to IBM. IBM may find us a work around sooner. 

In the meantime, I confirmed that our side installed MPI (currently used on Lassen) still works on rzansel with Flux. So our users should keep using that version until this problem is fixed.

**Comment by dongahn** [comment]
> In the meantime, I confirmed that our side installed MPI (currently used on Lassen) still works on rzansel with Flux. So our users should keep using that version until this problem is fixed.

A caveat is you now need to specify the old MPI library path ether  with `-Wl,-rpath` link flag or add that path to `LD_LIBRARY_PATH`.

```
mpixlc -g -O0 virtual_ring_mpi.c -o virtual_ring_mpi -Wl,-rpath=/usr/global/tools/flux/blueos_3_ppc64le_ib/spectrum-mpi-2019.06.24/lib
```

**Comment by dongahn** [comment]
Or 

```
mpixlc -g -O0 virtual_ring_mpi.c -o virtual_ring_mpi
bash-4.2$ jsrun -a 1 -c ALL_CPUS -g ALL_GPUS --bind=none -n 4 bash -c 'unset LD_PRELOAD; /usr/global/tools/flux/blueos_3_ppc64le_ib/flux-c0.18.0-s0.10.0/bin/flux start flux mini run -n 4 -o mpi=spectrum bash -c "LD_LIBRARY_PATH=/usr/global/tools/flux/blueos_3_ppc64le_ib/spectrum-mpi-2019.06.24/lib:$LD_LIBRARY_PATH; virtual_ring_mpi"'
[rzansel17:60758] mca_base_component_repository_open: unable to open mca_schizo_flux.so: File not found (ignored)
[rzansel17:60758] mca_base_component_repository_open: unable to open mca_pmix_flux.so: File not found (ignored)
[rzansel30:124194] mca_base_component_repository_open: unable to open mca_schizo_flux.so: File not found (ignored)
[rzansel52:111081] mca_base_component_repository_open: unable to open mca_schizo_flux.so: File not found (ignored)
[rzansel54:145086] mca_base_component_repository_open: unable to open mca_schizo_flux.so: File not found (ignored)
[rzansel30:124194] mca_base_component_repository_open: unable to open mca_pmix_flux.so: File not found (ignored)
[rzansel52:111081] mca_base_component_repository_open: unable to open mca_pmix_flux.so: File not found (ignored)
[rzansel54:145086] mca_base_component_repository_open: unable to open mca_pmix_flux.so: File not found (ignored)
size of this program is 4
rcvbuf: 3
```

Prepending to `LD_LIBRARY_PATH` within `spectrum-mpi/2019.06.24-flux` will be unlikely work though.

```
bash-4.2$ jsrun -a 1 -c ALL_CPUS -g ALL_GPUS --bind=none -n 4 bash -c "unset LD_PRELOAD; /usr/global/tools/flux/blueos_3_ppc64le_ib/flux-c0.18.0-s0.10.0/bin/flux start -v flux mini run -n 1 -o mpi=spectrum ldd virtual_ring_mpi"
/opt/ibm/spectrum_mpi/jsm_pmix/bin/stock/jsrun: symbol lookup error: /usr/global/tools/flux/blueos_3_ppc64le_ib/spectrum-mpi-2019.06.24/lib/pmix/mca_psec_native.so: undefined symbol: pmix_output_verbose
```

It prevents the top `jsrun` command from working correctly.

**Comment by SteVwonder** [comment]
> A caveat is you now need to specify the old MPI library path ether with -Wl,-rpath link flag or add that path to LD_LIBRARY_PATH.

Does that make the `mca_base_component_repository_open` warnings go away?

**Comment by dongahn** [comment]
> Does that make the mca_base_component_repository_open warnings go away?

It is still there. My hope is IBM's fix for this make that warnings go away. But I cannot test this because of the new problem: https://github.com/flux-framework/flux-core/issues/3095#issuecomment-666939302

**Comment by dongahn** [comment]
Convert this to discussion as well since this is a partner issue. Let me mark some W/R as answers as well.

**Comment by dongahn** [comment]
Bummer. You can only select one posting as "answer". So when we use this discussion feature, we may need a summary posting and select it as "answer."
